### PR TITLE
Test UserNotes\Sorter

### PR DESCRIPTION
**Pull request by cmb69**
We add a copy of the latest run-tests.php from the PHP-8.0 branch, and
tests for the `UserNotes\Sorter` with full code coverage.

### CODE REVIEW

**Comment by localheinz** [comment]
@cmb69 

How about running these tests on GitHub Actions?

See https://github.com/cmb69/web-php/pull/1 and https://github.com/localheinz/web-php/pull/3.

**Comment by cmb69** [comment]
> How about running these tests on GitHub Actions?

Yes, by all means! I didn't set up CI, because I saw PR #605, so wasn't quite sure.

**Comment by localheinz** [comment]
@cmb69 

Closed #605 and adjusted https://github.com/cmb69/web-php/pull/1!

**Comment by kamil-tekiela** [comment]
@cmb69 

I am getting test failures:
```
TEST 1/3 [D:\projects\web-php\tests\sort_notes_001.phpt]
========DIFF========
--
         ["total"]=>
         int(17)
         ["rating"]=>
015+     float(0.8235294117647058)
015-     float(0.82352941176471)
         ["sort"]=>
017+     float(87.41176470588235)
017-     float(87.411764705882)
       }
       [3]=>
       array(7) {
--
         ["total"]=>
         int(17)
         ["rating"]=>
032+     float(0.8235294117647058)
032-     float(0.82352941176471)
         ["sort"]=>
034+     float(87.41176470588235)
034-     float(87.411764705882)
       }
       [1]=>
       array(7) {
--
         ["rating"]=>
         float(0.5)
         ["sort"]=>
051+     float(45.49231350387337)
051-     float(45.492313503873)
       }
       [0]=>
       array(7) {
--
         float(43.4)
       }
     }
071+ 
072+ Termsig=-1073741819
========DONE========
FAIL sort some notes [D:\projects\web-php\tests\sort_notes_001.phpt]
TEST 2/3 [D:\projects\web-php\tests\sort_notes_002.phpt]
========DIFF========
     array(0) {
     }
003+ 
004+ Termsig=-1073741819
========DONE========
FAIL sort no notes [D:\projects\web-php\tests\sort_notes_002.phpt]
TEST 3/3 [D:\projects\web-php\tests\sort_notes_003.phpt]
========DIFF========
--
         float(41.4)
       }
     }
020+ 
021+ Termsig=-1073741819
========DONE========
FAIL sort a single note with no votes [D:\projects\web-php\tests\sort_notes_003.phpt]
=====================================================================
TIME END 2022-07-03 14:50:29
```
Any idea what I am doing wrong?

**Comment by localheinz** [comment]
@kamil-tekiela 

Which PHP version are you using?

**Comment by kamil-tekiela** [comment]
> PHP 8.0.3 (cli) (built: Mar  2 2021 23:34:05) ( ZTS Visual C++ 2019 x64 )

**Comment by localheinz** [comment]
@kamil-tekiela 

I'm getting similar test failures on 

- PHP 8.0.16
- PHP 8.1.3

Looks like something changed in regard to the precision.

**Comment by localheinz** [comment]
@cmb69 

Is there a way to run the tests without producing any files?

**Comment by localheinz** [comment]
@cmb69 

How do you feel about moving the tests?

```diff
diff --git a/tests/sort_notes_001.phpt b/tests/UserNotes/Sorter/sort_notes_001.phpt
similarity index 96%
rename from tests/sort_notes_001.phpt
rename to tests/UserNotes/Sorter/sort_notes_001.phpt
index adc07611..e9bfac6e 100644
--- a/tests/sort_notes_001.phpt
+++ b/tests/UserNotes/Sorter/sort_notes_001.phpt
@@ -2,7 +2,7 @@
 sort some notes
 --FILE--
 <?php
-require_once __DIR__ . "/../src/UserNotes/Sorter.php";
+require_once __DIR__ . "/../../src/UserNotes/Sorter.php";

 $notes = [
     [
diff --git a/tests/sort_notes_002.phpt b/tests/UserNotes/Sorter/sort_notes_002.phpt
similarity index 89%
rename from tests/sort_notes_002.phpt
rename to tests/UserNotes/Sorter/sort_notes_002.phpt
index def6d4b1..51387dcb 100644
--- a/tests/sort_notes_002.phpt
+++ b/tests/UserNotes/Sorter/sort_notes_002.phpt
@@ -2,7 +2,7 @@
 sort no notes
 --FILE--
 <?php
-require_once __DIR__ . "/../src/UserNotes/Sorter.php";
+require_once __DIR__ . "/../../src/UserNotes/Sorter.php";

 $notes = [];

diff --git a/tests/sort_notes_003.phpt b/tests/UserNotes/Sorter/sort_notes_003.phpt
similarity index 93%
rename from tests/sort_notes_003.phpt
rename to tests/UserNotes/Sorter/sort_notes_003.phpt
index 9e088464..c03af237 100644
--- a/tests/sort_notes_003.phpt
+++ b/tests/UserNotes/Sorter/sort_notes_003.phpt
@@ -2,7 +2,7 @@
 sort a single note with no votes
 --FILE--
 <?php
-require_once __DIR__ . "/../src/UserNotes/Sorter.php";
+require_once __DIR__ . "/../../src/UserNotes/Sorter.php";

 $notes = [
     [
```

**Comment by cmb69** [comment]
> Looks like something changed in regard to the precision.

Yeah, I actually expected this; `var_dump()` uses serialize_precision rather than precision as of PHP 8.0.0. This can be fixed by adding a respective entry to an `--INI--` section.

But I'm more concerned about

> 072+ Termsig=-1073741819

That is a segfault (aka. access violation on Windows). Maybe it is "just" the old revision (8.0.3).

> Is there a way to run the tests without producing any files on failure?

I don't think so, but there are

>    --temp-source <sdir>  --temp-target <tdir> [--temp-urlbase <url>]
                Write temporary files to <tdir> by replacing <sdir> from the
                filenames to generate with <tdir>. In general you want to make
                <sdir> the path to your source files and <tdir> some patch in
                your web page hierarchy with <url> pointing to <tdir>.

> How do you feel about moving the tests?

I think this is a good idea. The basenames could then be shorted: `sort_001.phpt` etc. Or maybe use more self explaining basenames instead: `sort_some.php`, `sort_no.phpt` and `sort_without_votes.phpt` or so.

**Comment by kamil-tekiela** [comment]
FWIW with 8.1.4 I only get precision errors and no segfaults. I don't have any other PHP 8.0 version installed at the moment, but maybe I will download a newer one and test again

**Comment by cmb69** [comment]
@kamil-tekiela, I just pushed a commit which is supposed to resolve the precision issue.

**Comment by kamil-tekiela** [comment]
PHP 7.4 - I get some completely different output.
PHP 8.0 - I get 3 segfaults
PHP 8.1 - All pass

File: .github/workflows/integrate.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# https://docs.github.com/en/actions
+
+name: "Integrate"
+
+on:
+  pull_request: null
```
Comment (cmb69) [inline .github/workflows/integrate.yaml:6]: @localheinz, does that mean that CI will not be triggered for PRs? IMO, we should run CI for PRs (every commit, etc.) There is already a guard for new contributors where CI has to be triggered manually.

**Comment by cmb69** [comment]
> PHP 7.4 - I get some completely different output.

The current tests pass for me with PHP 7.4.19 and 8.0.19 (haven't tested other versions).

**Comment by kamil-tekiela** [comment]
I downloaded 8.0.20 and no more segfaults. 

On PHP 7.4.9, the output is:
```
array(4) {
  [2] =>
  array(7) {
    'ts' =>
    int(1508179844)
    'upvotes' =>
    int(14)
    'downvotes' =>
    int(3)
    'score' =>
    int(11)
    'total' =>
    int(17)
    'rating' =>
    double(0.8)
    'sort' =>
    double(9.0E+1)
  }
  [3] =>
  array(7) {
    'ts' =>
    int(1508179844)
    'upvotes' =>
    int(14)
    'downvotes' =>
    int(3)
    'score' =>
    int(11)
    'total' =>
    int(17)
    'rating' =>
    double(0.8)
    'sort' =>
    double(9.0E+1)
  }
  [1] =>
  array(7) {
    'ts' =>
    int(1508180150)
    'upvotes' =>
    int(0)
    'downvotes' =>
    int(0)
    'score' =>
    int(0)
    'total' =>
    int(0)
    'rating' =>
    double(0.5)
    'sort' =>
    double(5.0E+1)
  }
  [0] =>
  array(7) {
    'ts' =>
    int(1613487094)
    'upvotes' =>
    int(0)
    'downvotes' =>
    int(2)
    'score' =>
    int(-2)
    'total' =>
    int(2)
    'rating' =>
    int(0)
    'sort' =>
    double(4.0E+1)
  }
}
```

File: .github/workflows/integrate.yaml (hunk 1 above)
Comment (localheinz) [reply .github/workflows/integrate.yaml:6]: No, this means that builds will run for pull requests and pushes to `master`.

Comment (cmb69) [reply .github/workflows/integrate.yaml:6]: Ah, thanks!

**Comment by cmb69** [comment]
> On PHP 7.4.9, the output is:

I get these results with `precision=1`, but that setting should be changed by the tests to `precision=-1`.

File: .github/workflows/integrate.yaml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+jobs:
+  tests:
+    name: "Tests"
+
+    runs-on: "ubuntu-18.04"
+
+    strategy:
+      matrix:
+        php-version:
+          - "7.2"
```
Comment (cmb69) [inline .github/workflows/integrate.yaml:20]: I just learned that we're using PHP 7.3.19 on that machine. Might make sense to update the version here.

Comment (localheinz) [reply .github/workflows/integrate.yaml:20]: Nice!

Will do as soon as I’m in front of a computer.

File: tests/sort_notes_001.phpt [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        "xwhen" => 1508179844,
+    ], [
+        "votes" => ["up" => 14, "down" => 3],
+        "xwhen" => 1508179844,
+    ],
+];
+
+$sorter = new phpweb\UserNotes\Sorter();
+$sorter->sort($notes);
+var_dump(array_map(function (array $note) {
```
Comment (localheinz) [inline tests/sort_notes_001.phpt:27]:
```suggestion
var_dump(array_map(function (array $note): array {
```

File: tests/sort_notes_002.phpt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+precision=-1
+--FILE--
+<?php
+require_once __DIR__ . "/../src/UserNotes/Sorter.php";
+
+$notes = [];
+
+$sorter = new phpweb\UserNotes\Sorter();
+$sorter->sort($notes);
+var_dump(array_map(function (array $note) {
```
Comment (localheinz) [inline tests/sort_notes_002.phpt:13]:
```suggestion
var_dump(array_map(function (array $note): array {
```

File: tests/sort_notes_003.phpt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+$notes = [
+    [
+        "votes" => ["up" => 0, "down" => 0],
+        "xwhen" => 1613487094,
+    ],
+];
+
+$sorter = new phpweb\UserNotes\Sorter();
+$sorter->sort($notes);
+var_dump(array_map(function (array $note) {
```
Comment (localheinz) [inline tests/sort_notes_003.phpt:18]:
```suggestion
var_dump(array_map(function (array $note): array {
```

Comment (cmb69) [reply tests/sort_notes_003.phpt]: Thank you @localheinz!

**Review by localheinz — COMMENTED** [review]
👍

**Review by kamil-tekiela — APPROVED** [review]
I don't know much about GH actions to judge if it's correct, but the rest looks fine to me.

**Comment by cmb69** [comment]
Well, then let's merge this; we can still improve the GH action integration if need be.

**Pull request merged**